resources/bundles/org.eclipse.core.resources/src/org/eclipse/core/internal/resources/AliasManager.java

```diff
@@ -624,27 +621,40 @@ public void updateAliases(IResource resource, IFileStore location, int depth, IP
 	/**
 	 * Process any structural changes that have occurred since the last alias
 	 * request.
+	 *
+	 * @return <code>true</code> if some structural changes where processed,
+	 *         <code>false</code> if no structural changes were detected
 	 */
-	private void updateStructureChanges() {
+	/*
+	 * This method is synchronized as it calls
+	 * buildLocationsMap/addToLocationsMap/addToLocationsMap which are not meant to
+	 * run in parallel. Incoming events/notifications that cause a structural change
+	 * can come from different threads, but we only want a single thread to process
+	 * them at a time.
+	 */
+	private synchronized boolean checkStructuralChanges() {
 		boolean hadChanges = false;
-		if (changedProjects) {
+		while (changedProjects) {
 			//if a project is added or removed, just recompute the whole world
 			changedProjects = false;
 			hadChanges = true;
 			buildLocationsMap();
-		} else {
-			//incrementally update location map for changed links
-			for (IResource resource : changedLinks) {
+		}
+		// incrementally update location map for changed links
+		while (!changedLinks.isEmpty()) {
```

Comment:
Don't use a while loop here otherwise addToLocationsMap could be called for same resource again.  which contradicts the use of a Set.

Reply:
It wouldn't really contradict: if a link is changed 2 times, and was already processed once in addLocationsToMap, and in the meantime it's changed again, then it needs to call addLocationsToMap again. The set is not really the set of IResource in general, but the Set of IResource than needs an update. It can happen that a resource needs several updates if it was changed while it was being processed.